client/src/components/Atoms/CodeInput.tsx

```diff
@@ -0,0 +1,114 @@
+import React, { useEffect } from 'react'
+import styled from 'styled-components'
+import { Classes } from '@blueprintjs/core'
+import { replaceAtIndex, range } from '../../utils/array'
+import { Row } from './Layout'
+import { assert } from '../utilities'
+
+const DigitInput = styled.input.attrs({
+  className: Classes.INPUT,
+  type: 'text',
+})`
+  text-align: center;
+`
+
+// We only support a small subset of the HTMLInputElement interface to start,
+// but this can be expanded as needed.
+type InputProps = Pick<
+  React.InputHTMLAttributes<HTMLInputElement>,
+  'name' | 'id'
+>
+
+interface ICodeInputProps extends InputProps {
+  length: number
+  value: string
+  onChange: (value: string) => void
+}
+
+/**
+ * CodeInput takes the same props as a controlled text input, but
+ * actually renders and coordinates multiple individual digit inputs.
+ *
+ * Example:
+ *  const [value, setValue] = useState('')
+ *  <CodeInput length={3} value={value} onChange={setValue} />
+ */
+const CodeInput: React.FC<ICodeInputProps> = ({
+  length,
+  value = '',
+  onChange,
+  ...htmlInputProps
+}) => {
+  assert(/^\d*$/.test(value), 'CodeInput value must be a string of digits')
```

Comment:
This isn't possible on the version of TS that we're on but just thought I'd note that, on the latest version of TS, you can use [template literal types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) to enforce this at compile time:

```ts
type NumericalString = `${number}`;
```

Kinda mind blowing!

<img width="285" alt="whoa-1" src="https://user-images.githubusercontent.com/12616928/198093960-cb4fe5e6-e79e-45b0-90d7-c0f957a0937e.png">

<img width="315" alt="whoa-2" src="https://user-images.githubusercontent.com/12616928/198093965-28e3b8ea-782d-4159-bc8d-bde58172e91d.png">

Reply:
That is pretty wild! Who comes up with this stuff??